..StatusBar get workspace properties like sid in :: PP

### Problem Description
I have an idea but not the necessary knowledge to carry it out, I have been trying for 48 hours.
the goal is to achieve that xmobars launched in multihead(+2 screens) configuration
show in their workspace pp list the sid or xrandr-name correspondig to ppCurrent and ppVisible, also the display corresponding to the xmobar instance, here is an example
1) with xrandr name
5(LVDS1)  *3(VGA1) 1(VIRTUAL1) 2 4 6 7 8 9 | VGA1 | TITLE.....
2) with sid
5(0)  *3(1) 1(2) 2 4 6 7 8 9 | 1 | TITLE.....
2) with independent screen (Not my case)
LVDS[ 1 *2 ] VGA1[ *1 2 ] VIRTUAL1[ 1 2 *3 ] | VGA1 | TITLE.......


### Configuration File

xmonad.hs
```haskell
import XMonad
import qualified Data.Set as S
import qualified Data.Map as M

import XMonad.Util.EZConfig
import XMonad.Util.Run

import XMonad.Hooks.StatusBar
import XMonad.Hooks.StatusBar.PP
import XMonad.Hooks.ManageDocks

main = xmonad $ docks $ dynamicSBs barSpawner def
       {
         terminal = "xterm"
       , startupHook = do
           spawn "xrandr --output VIRTUAL2 --off --output VIRTUAL1 --off --output LVDS1 --primary --mode 1024x768 --pos 0x1920 --rotate normal --output VGA1 --mode 1920x1080 --$
       }

xmobarCmd = "xmobar -B white -F black"

barSpawner :: ScreenId -> IO StatusBarConfig
barSpawner (S sid) = pure $
  statusBarPropTo ("_XMONAD_LOG_" ++ show sid) ( xmobarCmd ++ " --screen " ++ show sid ++ " ~/.xmobarrc" ++ show sid) (pure myXmobarPP)

myXmobarPP :: PP
myXmobarPP = def {
  ppCurrent          = xmobarColor "red"    "white"  . wrap " " " "
  ,ppVisible         = xmobarColor "gray"   "white"   . wrap " " " "
  ,ppTitle           = xmobarColor "black"  "white"   . shorten 128 . wrap "  " ""
  ,ppWsSep           = xmobarColor "white"  "white"   " "
  ,ppSep             = xmobarColor "black"  "white"   " | "
  }
```

~/.xmobarrc0
```haskell
Config {
  alpha              = 255
  , position         = Top
  , textOffset       = -1
  , iconOffset       = -1
  , lowerOnStart     = True
  , allDesktops      = False
  , overrideRedirect = False
  , pickBroadest     = True
  , persistent       = False
  , sepChar    = "%"
  , commands = [
      Run XPropertyLog "_XMONAD_LOG_0"
      , Run StdinReader
      , Run Date "%d/%m/%Y %H:%M:%S " "date" 10
      ]
  , alignSep   = "}{"
  , template = "%_XMONAD_LOG_0% }{ ..."
  }                                                                                                                                                                              


```

### Checklist

  - [x] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - I tested my configuration
    - [x] With `xmonad` version 0.17.0
    - [x] With `xmonad-contrib` 0.17

  with GHC 8.8.4
  build-depends:       
                     base >= 4.12
                     , containers
                     , data-default
                     , fuzzy
                     , monoid-subclasses
                     , mtl
                     , process
                     , unix
                     , split
                     , utf8-string
                     , X11 >= 1.9.2
                     , X11-xft >= 0.3.1
                     --, X11-extras >= 0.4
                     , xmobar >= 0.40
                     , xmonad >= 0.16 && < 0.18
                     , xmonad-contrib >= 0.16 && < 0.18
                     , text

> I think this is worthy of its own module, here's what I got:
> 
> ```haskell
> type WindowScreen
>   = W.Screen WorkspaceId (Layout Window) Window ScreenId ScreenDetail
> 
> type WorkspaceScreenCombiner = String -> WindowScreen -> String
> 
> monitorIds :: IO [(ScreenId, String)]
> monitorIds = return [(S 0, "HDMI-0"), (S 1, "eDP-1-1")]
> 
> 
> unscreen :: ScreenId -> Int
> unscreen (S sid) = sid
> 
> combineWithScreenNumber :: X WorkspaceScreenCombiner
> combineWithScreenNumber =
>   return $ \w sc -> w <> wrap "(" ")" (show . unscreen . W.screen $ sc)
> 
> combineWithScreenName :: X WorkspaceScreenCombiner
> combineWithScreenName = do
>   screenNames <- io monitorIds
>   return
>     $ \w sc -> w <> wrap "(" ")" (fromJust $ W.screen sc `lookup` screenNames)
> 
> renameWithScreen
>   :: X WorkspaceScreenCombiner -> X (String -> WindowSpace -> String)
> renameWithScreen xCombiner = do
>   combiner <- xCombiner
>   ss       <- withWindowSet (return . W.screens)
>   return $ \s w ->
>     maybe s (combiner s) (find ((== W.tag w) . W.tag . W.workspace) ss)
> 
> addScreen :: X WorkspaceScreenCombiner -> PP -> X PP
> addScreen xcombiner pp =
>   renameWithScreen xcombiner <&> \ren -> pp { ppRename = ppRename pp <=< ren }
> 
> 
> barSpawner :: ScreenId -> IO StatusBarConfig
> barSpawner (S sid) = pure $
>   statusBarPropTo ("_XMONAD_LOG_" ++ show sid) ( xmobarCmd ++ " --screen " ++ show sid ++ " ~/.xmobarrc" ++ show sid) myXmobarPP
> 
> myXmobarPP :: X PP
> myXmobarPP = 
>   addScreen combineWithScreenName $ def
>     { ppCurrent = xmobarColor "red" "white" . wrap " " " " . const c
>     , ppVisible = xmobarColor "gray" "white" . wrap " " " " . vs
>     , ppTitle   = xmobarColor "black" "white" . shorten 128 . wrap "  " ""
>     , ppWsSep   = xmobarColor "white" "white" " "
>     , ppSep     = xmobarColor "black" "white" " | "
>     }
> ```
> 
> You can replace `combineWithScreenName` with `combineWithScreenNumber` (I mocked `combineWithScreenName` since I didn't want to add `text` as a dependency)

Hi, friend i think the same and also works for get other parameters from workspaces in multihead scheme..

i'm getting this trying this setup :

xmonad.hs:65:30: error:
    • Variable not in scope:
        (<&>) :: X (String -> WindowSpace -> String) -> (t0 -> PP) -> X PP
    • Perhaps you meant one of these:
        ‘<>’ (imported from Prelude), ‘<&&>’ (imported from XMonad),
        ‘<+>’ (imported from XMonad)
   |
65 |   renameWithScreen xcombiner <&> \ren -> pp { ppRename = ppRename pp <=< ren }
   |                                                       ^^^

xmonad.hs:65:70: error:
    • Variable not in scope:
        (<=<)
          :: (String -> WindowSpace -> String)
             -> t0 -> String -> WindowSpace -> String
    • Perhaps you meant one of these:
        ‘=<<’ (imported from Prelude), ‘<=’ (imported from Prelude)
   |
65 |   renameWithScreen xcombiner <&> \ren -> pp { ppRename = ppRename pp <=< ren }
   |                                                                                                                               ^^^

xmonad.hs:81:68: error: Variable not in scope: c :: String
   |
81 |     { ppCurrent = xmobarColor "red" "white" . wrap " " " " . const c
   |                                                                                                        ^

xmonad.hs:82:63: error:
    Variable not in scope: vs :: WorkspaceId -> String
   |
82 |     , ppVisible = xmobarColor "gray" "white" . wrap " " " " . vs
   |                                                                                              ^^
Linking /home/username/.xmonad/dist-newstyle/build/i386-linux/ghc-8.8.4/my-xmonad-0.1.0.0/x/xmonadctl/build/xmonadctl/xmonadctl ...
cabal: Failed to build exe:xmonad from my-xmonad-0.1.0.0.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -121,6 +121,9 @@
     - Added `keymaskToString` and `keyToString` to show a key mask and a
       key in the style of `XMonad.Util.EZConfig`.
 
+    - Added `WindowScreen`, which is a type synonym for the specialized `Screen`
+      type, that results from the `WindowSet` definition in `XMonad.Core`.
+
   * `XMonad.Util.XUtils`
 
     - Added `withSimpleWindow`, `showSimpleWindow`, `WindowConfig`, and

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -40,6 +40,12 @@
     A wrapper around the 'ReadP' parser combinator, providing behaviour
     that's closer to the more popular parser combinator libraries.
 
+  * `XMonad.Hooks.StatusBar.WorkspaceScreen`
+
+    In multi-head setup, it might be useful to have screen information of the
+    visible workspaces combined with the workspace name, for example in a status
+    bar. This module provides utility functions to do just that.
+
 ### Bug Fixes and Minor Changes
 
   * `XMonad.Prompt.OrgMode`

--- a/XMonad/Actions/EasyMotion.hs
+++ b/XMonad/Actions/EasyMotion.hs
@@ -271,7 +271,7 @@ handleSelectWindow c = do
         $ M.elems
         $ M.mapWithKey (\sid ks -> buildOverlays ks <$> sortedOverlayWindows sid) m
      where
-      screenById :: ScreenId -> Maybe (W.Screen WorkspaceId (Layout Window) Window ScreenId ScreenDetail)
+      screenById :: ScreenId -> Maybe WindowScreen
       screenById sid = find ((== sid) . W.screen) (W.screens ws)
       visibleWindowsOnScreen :: ScreenId -> [Window]
       visibleWindowsOnScreen sid = filter (`elem` toList mappedWins) $ W.integrate' $ screenById sid >>= W.stack . W.workspace

--- a/XMonad/Actions/Navigation2D.hs
+++ b/XMonad/Actions/Navigation2D.hs
@@ -388,7 +388,7 @@ data Navigation2DConfig = Navigation2DConfig
   }
 
 -- | Shorthand for the tedious screen type
-type Screen = W.Screen WorkspaceId (Layout Window) Window ScreenId ScreenDetail
+type Screen = WindowScreen
 
 -- | Convenience function for enabling Navigation2D with typical keybindings.
 -- Takes a Navigation2DConfig, an (up, left, down, right) tuple, a mapping from

--- a/XMonad/Hooks/StatusBar/WorkspaceScreen.hs
+++ b/XMonad/Hooks/StatusBar/WorkspaceScreen.hs
@@ -1,0 +1,105 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeApplications #-}
+
+{- |
+ Module      :  XMonad.Hooks.StatusBar.WorkspaceScreen
+ Description :  Combine workspace names with screen information
+ Copyright   :  (c) Yecine Megdiche <yecine.megdiche@gmail.com>
+ License     :  BSD3-style (see LICENSE)
+
+ Maintainer  :  Yecine Megdiche <yecine.megdiche@gmail.com>
+ Stability   :  unstable
+ Portability :  unportable
+
+ In multi-head setup, it might be useful to have screen information of the
+ visible workspaces combined with the workspace name, for example in a status
+ bar. This module provides utility functions to do just that.
+-}
+module XMonad.Hooks.StatusBar.WorkspaceScreen
+    (
+    -- * Usage
+    -- $usage
+      combineWithScreen
+    , combineWithScreenName
+    , combineWithScreenNumber
+    , WorkspaceScreenCombiner
+    -- * Limitations
+    -- $limitations
+    ) where
+
+import           Graphics.X11.Xrandr
+import           XMonad
+import           XMonad.Hooks.StatusBar.PP
+import           XMonad.Prelude
+import qualified XMonad.StackSet               as W
+
+{- $usage
+ You can use this module with the following in your @~\/.xmonad\/xmonad.hs@:
+
+ > import XMonad
+ > import XMonad.Hooks.StatusBar
+ > import XMonad.Hooks.StatusBar.PP
+ > import XMonad.Hooks.StatusBar.WorkspaceScreen
+
+ For example, to add the screen number in parentheses to each visible
+ workspace number, you can use 'combineWithScreenNumber':
+
+ > myWorkspaceScreenCombiner :: WorkspaceId -> String -> String
+ > myWorkspaceScreenCombiner w sc = w <> wrap "(" ")" sc
+ >
+ > mySB = statusBarProp "xmobar" (combineWithScreenNumber myWorkspaceScreenCombiner xmobarPP)
+ > main = xmonad $ withEasySB mySB defToggleStrutsKey def
+
+ This will annotate the workspace names as following:
+
+ > [1(0)] 2 3 4 <5(1)> 6 7 8 9
+
+ To use the screen's name instead, checkout 'combineWithScreenName':
+
+ > [1(eDP-1)] 2 3 4 <5(HDMI-1)> 6 7 8 9
+
+ For advanced cases, use 'combineWithScreen'.
+-}
+
+{- $limitations
+ For simplicity, this module assumes xmonad screen ids match screen/monitor
+ numbers as managed by the X server (for example, as given by @xrandr
+ --listactivemonitors@). Thus, it may not work well when screens show an
+ overlapping range of the framebuffer, e.g. when using a projector. This also
+ means that it doesn't work with "XMonad.Layout.LayoutScreens".
+ (This isn't difficult to fix, PRs welcome.)
+-}
+
+-- | Type synonym for a function that combines a workspace name with a screen.
+type WorkspaceScreenCombiner = WorkspaceId -> WindowScreen -> String
+
+-- | A helper function that returns a list of screen names.
+screenNames :: X [Maybe String]
+screenNames = do
+    XConf { display, theRoot } <- ask
+    let getName mi = getAtomName display (xrr_moninf_name mi)
+    io
+        $   maybe (pure []) (traverse getName)
+        =<< xrrGetMonitors display theRoot True
+
+-- | Combine a workspace name with the screen name it's visible on.
+combineWithScreenName :: (WorkspaceId -> String -> String) -> PP -> X PP
+combineWithScreenName c = combineWithScreen $ do
+    screens <- screenNames
+    return $ \w sc -> maybe w (c w) $ join (screens !? fi (W.screen sc))
+
+-- | Combine a workspace name with the screen number it's visible on.
+combineWithScreenNumber :: (WorkspaceId -> String -> String) -> PP -> X PP
+combineWithScreenNumber c =
+    combineWithScreen . return $ \w sc -> c w (show @Int . fi . W.screen $ sc)
+
+-- | Combine a workspace name with a screen according to the given
+-- 'WorkspaceScreenCombiner'.
+combineWithScreen :: X WorkspaceScreenCombiner -> PP -> X PP
+combineWithScreen xCombiner pp = do
+    combiner <- xCombiner
+    ss       <- withWindowSet (return . W.screens)
+    return $ pp
+        { ppRename = ppRename pp <=< \s w ->
+            maybe s (combiner s) (find ((== W.tag w) . W.tag . W.workspace) ss)
+        }

--- a/XMonad/Layout/IndependentScreens.hs
+++ b/XMonad/Layout/IndependentScreens.hs
@@ -148,7 +148,7 @@ withWspOnScreen screenId operation ws = case workspaceOnScreen screenId ws of
     Nothing -> ws
 
 -- | Get the workspace that is active on a given screen.
-screenOnMonitor :: ScreenId -> WindowSet -> Maybe (W.Screen WorkspaceId (Layout Window) Window ScreenId ScreenDetail)
+screenOnMonitor :: ScreenId -> WindowSet -> Maybe WindowScreen
 screenOnMonitor screenId ws = find ((screenId ==) . W.screen) (W.current ws : W.visible ws)
 
 -- | Focus a window, switching workspace on the correct Xinerama screen if neccessary.

--- a/XMonad/Prelude.hs
+++ b/XMonad/Prelude.hs
@@ -32,6 +32,7 @@ module XMonad.Prelude (
     specialKeys,
     multimediaKeys,
     functionKeys,
+    WindowScreen,
 ) where
 
 import Foreign (alloca, peek)
@@ -57,6 +58,7 @@ import Data.Bits
 import Data.List.NonEmpty (NonEmpty ((:|)))
 import Data.Tuple (swap)
 import GHC.Stack
+import qualified XMonad.StackSet as W
 
 -- | Short for 'fromIntegral'.
 fi :: (Integral a, Num b) => a -> b
@@ -415,3 +417,7 @@ multimediaKeys = filter ((/= noSymbol) . snd) . map (id &&& stringToKeysym) $
   , "XF86_Prev_VMode"
   , "XF86Bluetooth"
   ]
+
+-- | The specialized 'W.Screen' derived from 'WindowSet'.
+type WindowScreen -- FIXME move to core
+    = W.Screen WorkspaceId (Layout Window) Window ScreenId ScreenDetail

--- a/XMonad/Util/Loggers.hs
+++ b/XMonad/Util/Loggers.hs
@@ -48,7 +48,7 @@ module XMonad.Util.Loggers (
 
   ) where
 
-import XMonad (liftIO, Window, gets)
+import XMonad (liftIO, gets)
 import XMonad.Core
 import qualified XMonad.StackSet as W
 import XMonad.Hooks.StatusBar.PP
@@ -56,7 +56,7 @@ import XMonad.Util.Font (Align (..))
 import XMonad.Util.NamedWindows (getName)
 
 import Control.Exception as E
-import XMonad.Prelude (find, fromMaybe, isPrefixOf, isSuffixOf)
+import XMonad.Prelude (find, fromMaybe, isPrefixOf, isSuffixOf, WindowScreen)
 import Data.Time (defaultTimeLocale, formatTime, getCurrentTime)
 import System.Directory (getDirectoryContents)
 import System.IO (hGetLine)
@@ -271,9 +271,6 @@ logCurrentOnScreen = withScreen $ logConst . W.tag . W.workspace
 logLayoutOnScreen :: ScreenId -> Logger
 logLayoutOnScreen =
   withScreen $ logConst . description . W.layout . W.workspace
-
--- | A shortcut to a screen
-type WindowScreen = W.Screen WorkspaceId (Layout Window) Window ScreenId ScreenDetail
 
 -- | A helper function to create screen-specific loggers.
 withScreen :: (WindowScreen -> Logger) -> ScreenId -> Logger

--- a/XMonad/Util/PureX.hs
+++ b/XMonad/Util/PureX.hs
@@ -53,7 +53,7 @@ module XMonad.Util.PureX (
 
 -- xmonad
 import XMonad
-import XMonad.Prelude (Any (..), liftA2)
+import XMonad.Prelude (Any (..), liftA2, WindowScreen)
 import qualified XMonad.StackSet as W
 import qualified XMonad.Actions.FocusNth
 
@@ -220,9 +220,7 @@ peek :: XLike m => m (Maybe Window)
 peek = withWindowSet' (return . W.peek)
 
 -- | Get the current screen.
-curScreen
-  :: XLike m
-  => m (W.Screen WorkspaceId (Layout Window) Window ScreenId ScreenDetail)
+curScreen :: XLike m => m WindowScreen
 curScreen = withWindowSet' (return . W.current)
 
 -- | Get the current workspace.

--- a/xmonad-contrib.cabal
+++ b/xmonad-contrib.cabal
@@ -200,6 +200,7 @@ library
                         XMonad.Hooks.SetWMName
                         XMonad.Hooks.StatusBar
                         XMonad.Hooks.StatusBar.PP
+                        XMonad.Hooks.StatusBar.WorkspaceScreen
                         XMonad.Hooks.TaffybarPagerHints
                         XMonad.Hooks.ToggleHook
                         XMonad.Hooks.UrgencyHook

